File: LeetCodeProblemsGo/LeetCode1976NumberOfWaysToArriveAtDestination.go
package main

import (
	"container/heap"
	"math"
)

func countPaths(n int, roads [][]int) int {
	distances := make([]int64, n)
	for i := range distances {
		distances[i] = math.MaxInt
	}
	distances[0] = 0
	adjPlus := make(map[int]map[int]int)
	for i := 0; i < n; i++ {
		adjPlus[i] = make(map[int]int)
	}
	for _, edge := range roads {
		adjPlus[edge[0]][edge[1]] = edge[2]
		adjPlus[edge[1]][edge[0]] = edge[2]
	}
	minHeap := &MinHeapDistanceNodeItem{}
	heap.Init(minHeap)
	heap.Push(minHeap, &DistanceNodeItem{distance: 0, node: 0})
	var newDistance int64
	for minHeap.Len() > 0 {
		nodeInfo := heap.Pop(minHeap).(*DistanceNodeItem)
		if nodeInfo.distance > distances[nodeInfo.node] {
			continue
		}
		for key, value := range adjPlus[nodeInfo.node] {
			newDistance = distances[nodeInfo.node] + int64(value)
			if newDistance >= distances[key] {
				continue
			}
			distances[key] = newDistance
			heap.Push(minHeap, &DistanceNodeItem{distance: newDistance, node: key})
		}
	}
	adjMinus := make(map[int][]int)
	for _, edge := range roads {
		if distances[edge[0]]+int64(edge[2]) == distances[edge[1]] {
			if _, exists := adjMinus[edge[1]]; !exists {
				adjMinus[edge[1]] = []int{}
			}
			adjMinus[edge[1]] = append(adjMinus[edge[1]], edge[0])
		}
		if distances[edge[1]]+int64(edge[2]) == distances[edge[0]] {
			if _, exists := adjMinus[edge[0]]; !exists {
				adjMinus[edge[0]] = []int{}
			}
			adjMinus[edge[0]] = append(adjMinus[edge[0]], edge[1])
		}
	}
	memoisation := make(map[int]int)
	memoisation[0] = 1
	return CalculateWayCounts(adjMinus, memoisation, n-1)
}
func CalculateWayCounts(adjMinus map[int][]int, memoisation map[int]int, node int) int {
	if value, exists := memoisation[node]; exists {
		return value
	}
	var result int64
	if _, exists := adjMinus[node]; !exists {
		return 0
	}
	for _, vert := range adjMinus[node] {
		result += int64(CalculateWayCounts(adjMinus, memoisation, vert))
		result %= 1_000_000_007
	}
	memoisation[node] = int(result % 1_000_000_007)
	return memoisation[node]
}

type DistanceNodeItem struct {
	distance int64
	node     int
	index    int
}

type MinHeapDistanceNodeItem []*DistanceNodeItem

func (minHeap MinHeapDistanceNodeItem) Len() int {
	return len(minHeap)
}
func (minHeap MinHeapDistanceNodeItem) Less(i, j int) bool {
	return minHeap[i].distance < minHeap[j].distance
}
func (minHeap MinHeapDistanceNodeItem) Swap(i, j int) {
	minHeap[i], minHeap[j] = minHeap[j], minHeap[i]
}

func (minHeap *MinHeapDistanceNodeItem) Push(x any) {
	length := len(*minHeap)
	item := x.(*DistanceNodeItem)
	item.index = length
	*minHeap = append(*minHeap, item)
}

func (minHeap *MinHeapDistanceNodeItem) Pop() any {
	old := *minHeap
	length := len(*minHeap)
	item := old[length-1]
	old[length-1] = nil
	item.index = -1
	*minHeap = old[:length-1]
	return item
}

File: LeetCodeProblems/Problems/LeetCode1976NumberOfWaysToArriveAtDestination.cs
using System.Text.Json;

namespace LeetCodeProblems.Problems;

public class LeetCode1976NumberOfWaysToArriveAtDestination : BaseProblemClass
{
    public int CountPaths(int n, int[][] roads)
    {
        var distances = Enumerable.Repeat(long.MaxValue, n).ToArray();
        distances[0] = 0;
        var visited = new bool[n];
        Dictionary<int, List<(int node,int distance)>> adjPlus = [];
        for (int i = 0; i < n; i++)
            adjPlus[i] = [];
        foreach (var edge in roads)
        {
            adjPlus[edge[0]].Add((edge[1], edge[2]));
            adjPlus[edge[1]].Add((edge[0], edge[2]));
        }

        var queue = new PriorityQueue<(long distance, int node), long>();
        queue.Enqueue((0, 0), 0);
        long newDistance = 0;
        while (queue.Count > 0)
        {
            var nodeInfo = queue.Dequeue();
            if(nodeInfo.distance >distances[nodeInfo.node])
                continue;
            foreach (var edgeInfo in adjPlus[nodeInfo.node])
            {
                newDistance = distances[nodeInfo.node] + edgeInfo.distance;
                if(newDistance >= distances[edgeInfo.node])
                    continue;
                distances[edgeInfo.node] = newDistance;
                queue.Enqueue((newDistance, edgeInfo.node), newDistance);
            }
        }

        Dictionary<int, HashSet<int>> adjMinus = [];
        foreach (var edge in roads)
        {
            if (distances[edge[0]] + edge[2] == distances[edge[1]])
            {
                if(!adjMinus.ContainsKey(edge[1]))
                    adjMinus[edge[1]] = [];
                adjMinus[edge[1]].Add(edge[0]);
            }

            if (distances[edge[1]] + edge[2] == distances[edge[0]])
            {
                if(!adjMinus.ContainsKey(edge[0]))
                    adjMinus[edge[0]] = [];
                adjMinus[edge[0]].Add(edge[1]);
            }
        }

        Dictionary<int, int> memoisation = [];
        memoisation[0] = 1;
        return CalculateWayCounts(adjMinus, memoisation, n - 1);
    }

    int CalculateWayCounts(Dictionary<int, HashSet<int>> adjMinus, Dictionary<int,int> memoisation,int node)
    {
        if(memoisation.ContainsKey(node))
            return memoisation[node];
        long result = 0;
        if(!adjMinus.ContainsKey(node))
            return 0;
        foreach (var vert in adjMinus[node])
        {
            result += CalculateWayCounts(adjMinus,memoisation, vert);
            result %= 1_000_000_007;
        }
        memoisation[node]= (int)(result% 1_000_000_007);
        return memoisation[node];
    }
    public override void Run()
    {
        int n = 7;
        int[][] roads =
        [
            [0, 6, 7], [0, 1, 2], [1, 2, 3], [1, 3, 3], [6, 3, 3], [3, 5, 1], [6, 5, 1], [2, 5, 1], [0, 4, 5], [4, 6, 2]
        ];
        Console.WriteLine(CountPaths(n, roads));
    }
}